### ISSUE TITLE
S3885 gives incorrect non-working advice 

### ISSUE THREAD

**Issue by StingyJack** [SonarSource/sonar-dotnet#4327]
### Description

I have code thats loading an assembly the user directs from disk that then does stuff. This requires use of `Assembly.LoadFrom(string pathToAssembly)` as that is how to specify a file system path. SonarLint is telling me I should use `Assembly.Load(string assemblyString)` instead, but that doesn't accept a path, it expects an assembly name. That simply wont work.

Here is a visual depiction of the difference
``` csharp
Assembly.LoadFrom(@"c:\temp\Toolbox.dll");
Assembly.Load("Toolbox.dll, version=1.1.10.1220, locale=en, publickey=1234567890123456789012345678901234567890")
```

The [RSPEC ](https://jira.sonarsource.com/browse/RSPEC-3885)could not be any more incorrect in this scenario. 
> Use another method, and you might end up with a dll other than the one you expected.

If I am loading the assembly from a specific path on disk, then I am loading the dll that I expected. If I use `Assembly.Load(string assemblyName)`, the runtime is going to load the first assembly that matches the name spec given, which may not actually be the one I expect. It could be pulling from the GAC or another folder entirely from where the one is that I want to load. **It** has the chance of loading an assembly I didnt expect, not `LoadFrom` or `LoadFile`.

`LoadFile` and `LoadFrom` are really no different from the `Assembly.Load(byte[] assemblyBytes, )` overloads, which also cant load an unexpected assembly. This rule is fine for `LoadWithPartialName` but because that is being deprecated, not because it will load an unexpected assembly.

### Expected behavior

I expect it to not report this usage as incorrect. 
I also expect it to not give non-working advice

### Actual behavior

It does the opposite of both expectations

**Comment by csaba-sagi-sonarsource** [comment]
Hi @StingyJack 

There is a comparision of the two methods in the following  [blog post](https://docs.microsoft.com/en-us/archive/blogs/suzcook/loadfile-vs-loadfrom), so please take a look at it.

Also with LoadFrom you will load the assembly from the path you defined but It can be replaced on the disk, so that is how you can end up with something you did not expect.
With Load you can define the exact version and publickey and you will get what you asked for, even if it was pulled from the GAC or another folder entirely. It will be that assembly with that version and publickey.

**Comment by StingyJack** [comment]
That blog post is so out of date its not very reliable as a source. An assembly can be replaced on disk or otherwise intercepted for either scenario.  Its only a little more difficult for a named assembly because an SNK file is needed, [but its definitely possible](https://blog.fredrikhaglund.se/blog/2009/02/06/how-to-patchfixhack-someone-elses-assembly/) (had to use this ~ 10 years ago because the source code for one of our assemblies was corrupted). 

Then again if someone has access to the filesystem to replace an assembly, that system is already compromised. 

https://docs.microsoft.com/en-us/dotnet/framework/deployment/best-practices-for-assembly-loading  would be a better source for advice than that blog post. 

The point I was making is that LoadFrom is going to load the assembly I have directed it to load, not the first assembly matching the name given to it from any of the probing paths. . There is no difference between `LoadFrom(string)` and `Load(byte[])` as both load an arbitrary assembly, yet `Load(byte[])` doesnt get a S3885

**Comment by csaba-sagi-sonarsource** [comment]
Thank you @StingyJack  for the assembly loading best practices link.

After reading I agree with you that we can still improve the rule a bit, like reporting issues for the Load overloads that do not load into the default load context.

However I would still not say that the current rule gives incorrect non-working advice. The link you provided also suggests the following:

- We recommend that you switch to the Assembly.Load method and supply full assembly display names to ensure that correct versions are always used.
- If an assembly with the same identity is already loaded in the load-from context, LoadFrom returns the loaded assembly even if a different path was specified.

So, I suggest to close this issue and open another one to improve the rule. By covering overloads of the Load method that are non compliant and also checking if the full assembly display name was used.

**Comment by StingyJack** [comment]
> to ensure that correct versions are always used.
But I'm not trying to load a specific version, I'm trying to load a specific file, which is most often the case when loading an assembly. I've made a lot of stuff that loads assemblies and not a single one has wanted to use a FQAN to load.  I dont think I've come across any programs that did an explicit load by FQAN that werent MSFT .net tools.

The rule is not applying the scope and context appropriately by assuming all usages are going to be of a specific use case. 

Also I am pretty sure that assembly resolvers operate on a file path basis

**Comment by csaba-sagi-sonarsource** [comment]
We’ve had an internal talk. And we agree that for some cases the rule is not applicable, yet it is aligned with the Microsoft best practices for assembly loading. So, we will not change the rule at this point and I will close this issue.

**Comment by StingyJack** [comment]
>   And we agree that for some cases the rule is not applicable, yet it is aligned with the Microsoft best practices for assembly loading

Conflicting statement. If the rule were aligned, there would be no non-applicable cases. 

"Consider" switching to the default load context instead of the load-from (LoadFile, LoadFrom) context. There is not an "Avoid" on that.  Even MS recognizes that the LoadFrom context is necessary. Your rule should as well and not flag every usage of LoadFile or LoadFrom as a problem.

**Comment by tzdlr** [comment]
not just conflicting, in my case im loading a dll from a very specific path. Yes, i could change everything to pass in byte[] you cant assume evrything is properly located...

**Comment by manfred-brands** [comment]
In my case I'm calling the `Assembly.LoadFrom` inside an `AssemblyResolve` handler because the system cannot find the file by assembly name. Using `Assembly.Load` would be futile and likely result in a stack overflow.

**Comment by pavel-mikula-sonarsource** [comment]
@manfred-brands Please create a new issue from this, with full context and some snippet. We might be able to stop raising issues inside that `AssemblyResolve` method (if the invocation is directly inside that method)

**Comment by NETSphereSoft** [comment]
> [@manfred-brands](https://github.com/manfred-brands) Please create a new issue from this, with full context and some snippet. We might be able to stop raising issues inside that `AssemblyResolve` method (if the invocation is directly inside that method)

We have the same problem, is there am issue already?

**Comment by pavel-mikula-sonarsource** [comment]
@NETSphereSoft We don't have any other S3885 issue opened, so there wasn't one created.

For the moment, please report a new one here: https://community.sonarsource.com/tags/c/clean-code/fp/7/dotnet

Opening new GH issues will be available later.

**Comment by NETSphereSoft** [comment]
It's done here already:

https://community.sonarsource.com/t/csharp-3885-no-solution-possible-when-used-in-assemblyresolve-event/143132
